src/lang/pretty_print.rs

```diff
@@ -0,0 +1,669 @@
+//! Pretty-printing functionality for POD2 custom predicates
+
+use std::fmt::Write;
+
+use hex::ToHex;
+use itertools::Itertools;
+
+use crate::middleware::{
+    CustomPredicate, CustomPredicateBatch, NativePredicate, Predicate, StatementTmpl,
+    StatementTmplArg, TypedValue, Value,
+};
+
+/// Trait for converting AST nodes to Podlang source code
+///
+/// This trait provides a consistent interface for pretty-printing different
+/// types of AST nodes back to their Podlang source representation.
+pub trait PrettyPrint {
```

Comment:
Have you considered changing the API of the trait methods to receive a mutable `Writer` instead of returning a `String`?  Like this:
```rust
fn fmt_podlang_with_indent(&self, w: &mut dyn fmt::Write, indent: usize) -> Result<(), fmt::Error>;
```

Then you don't need to allocate a string for each implementation and you can just continuously write to the `w` (as a stream)

Reply:
I see you're already using the `write!` macro extensively, those calls would look the same but instead of writing to `&mut result` you'd write to `w`.
And whenever you have `format!(...)` you''ll have to change it for `write!(w, ...)`